File: src/test/java/net/pterodactylus/sone/web/CreatePostPageTest.java
package net.pterodactylus.sone.web;

import static net.pterodactylus.util.web.Method.POST;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import net.pterodactylus.sone.data.Sone;

import com.google.common.base.Optional;
import org.junit.Test;

/**
 * Unit test for {@link CreatePostPage}.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class CreatePostPageTest extends WebPageTest {

	private final CreatePostPage page = new CreatePostPage(template, webInterface);

	@Test
	public void pageReturnsCorrectPath() {
		assertThat(page.getPath(), is("createPost.html"));
	}

	@Test
	public void returnPageIsSetInTemplateContext() throws Exception {
		addHttpRequestParameter("returnPage", "returnPage.html");
		page.processTemplate(freenetRequest, templateContext);
		assertThat(templateContext.get("returnPage"), is((Object) "returnPage.html"));
	}

	@Test
	public void postIsCreatedCorrectly() throws Exception {
		addHttpRequestParameter("returnPage", "returnPage.html");
		addHttpRequestParameter("text", "post text");
		request("", POST);
		expectedException.expect(WebTestUtils.redirectsTo("returnPage.html"));
		try {
			page.processTemplate(freenetRequest, templateContext);
		} finally {
			verify(core).createPost(currentSone, Optional.<Sone>absent(), "post text");
		}
	}

	@Test
	public void creatingAnEmptyPostIsDenied() throws Exception {
		addHttpRequestParameter("returnPage", "returnPage.html");
		addHttpRequestParameter("text", "   ");
		request("", POST);
		page.processTemplate(freenetRequest, templateContext);
		assertThat(templateContext.get("errorTextEmpty"), is((Object) true));
	}

	@Test
	public void aSenderCanBeSelected() throws Exception {
		addHttpRequestParameter("returnPage", "returnPage.html");
		addHttpRequestParameter("text", "post text");
		addHttpRequestParameter("sender", "sender-id");
		Sone sender = mock(Sone.class);
		addLocalSone("sender-id", sender);
		request("", POST);
		expectedException.expect(WebTestUtils.redirectsTo("returnPage.html"));
		try {
			page.processTemplate(freenetRequest, templateContext);
		} finally {
			verify(core).createPost(sender, Optional.<Sone>absent(), "post text");
		}
	}

	@Test
	public void aRecipientCanBeSelected() throws Exception {
		addHttpRequestParameter("returnPage", "returnPage.html");
		addHttpRequestParameter("text", "post text");
		addHttpRequestParameter("recipient", "recipient-id");
		Sone recipient = mock(Sone.class);
		addSone("recipient-id", recipient);
		request("", POST);
		expectedException.expect(WebTestUtils.redirectsTo("returnPage.html"));
		try {
			page.processTemplate(freenetRequest, templateContext);
		} finally {
			verify(core).createPost(currentSone, Optional.of(recipient), "post text");
		}
	}

}

File: src/test/kotlin/net/pterodactylus/sone/web/CreatePostPageTest.kt
package net.pterodactylus.sone.web

import com.google.common.base.Optional.absent
import net.pterodactylus.sone.data.Sone
import net.pterodactylus.sone.test.asOptional
import net.pterodactylus.sone.test.mock
import net.pterodactylus.util.web.Method.POST
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Test
import org.mockito.Mockito.verify

/**
 * Unit test for [CreatePostPage].
 */
class CreatePostPageTest: WebPageTest() {

	private val page = CreatePostPage(template, webInterface)

	override fun getPage() = page

	@Test
	fun `page returns correct path`() {
		assertThat(page.path, equalTo("createPost.html"))
	}

	@Test
	fun `page requires login`() {
		assertThat(page.requiresLogin(), equalTo(true))
	}

	@Test
	fun `return page is set in template context`() {
		addHttpRequestParameter("returnPage", "return.html")
		page.processTemplate(freenetRequest, templateContext)
		assertThat(templateContext["returnPage"], equalTo<Any>("return.html"))
	}

	@Test
	fun `post is created correctly`() {
		request("", POST)
		addHttpRequestParameter("returnPage", "return.html")
		addHttpRequestParameter("text", "post text")
		verifyRedirect("return.html") {
			verify(core).createPost(currentSone, absent(), "post text")
		}
	}

	@Test
	fun `creating an empty post is denied`() {
		request("", POST)
		addHttpRequestParameter("returnPage", "return.html")
		addHttpRequestParameter("text", "  ")
		page.processTemplate(freenetRequest, templateContext)
		assertThat(templateContext["errorTextEmpty"], equalTo<Any>(true))
	}

	@Test
	fun `a sender can be selected`() {
		request("", POST)
		addHttpRequestParameter("returnPage", "return.html")
		addHttpRequestParameter("text", "post text")
		addHttpRequestParameter("sender", "sender-id")
		val sender = mock<Sone>()
		addLocalSone("sender-id", sender)
		verifyRedirect("return.html") {
			verify(core).createPost(sender, absent(), "post text")
		}
	}

	@Test
	fun `a recipient can be selected`() {
		request("", POST)
		addHttpRequestParameter("returnPage", "return.html")
		addHttpRequestParameter("text", "post text")
		addHttpRequestParameter("recipient", "recipient-id")
		val recipient = mock<Sone>()
		addSone("recipient-id", recipient)
		verifyRedirect("return.html") {
			verify(core).createPost(currentSone, recipient.asOptional(), "post text")
		}
	}

	@Test
	fun `text is filtered correctly`() {
		request("", POST)
		addHttpRequestParameter("returnPage", "return.html")
		addHttpRequestParameter("text", "post http://localhost:12345/KSK@foo text")
		addHttpRequestHeader("Host", "localhost:12345")
		verifyRedirect("return.html") {
			verify(core).createPost(currentSone, absent(), "post KSK@foo text")
		}
	}

}
